R-package/vignettes/CharRnnModel.Rmd

```diff
@@ -36,237 +32,262 @@ download.data <- function(data_dir) {
+X.train.label <- Y[, 1:as.integer(samples * train.val.fraction)]
+X.val.label <- Y[, -(1:as.integer(samples * train.val.fraction))]
 
-```{r}
-model <- mx.lstm(X.train, X.val, 
-                 ctx=mx.cpu(),
-                 num.round=num.round, 
-                 update.period=update.period,
-                 num.lstm.layer=num.lstm.layer, 
-                 seq.len=seq.len,
-                 num.hidden=num.hidden, 
-                 num.embed=num.embed, 
-                 num.label=vocab,
-                 batch.size=batch.size, 
-                 input.size=vocab,
-                 initializer=mx.init.uniform(0.1), 
-                 learning.rate=learning.rate,
-                 wd=wd,
-                 clip_gradient=clip_gradient)
+train_buckets <- list("100"=list(data=X.train.data, label=X.train.label))
+eval_buckets <- list("100"=list(data=X.val.data, label=X.val.label))
 
+train_buckets <- list(buckets = train_buckets, dic = dic, rev_dic = rev_dic)
+eval_buckets <- list(buckets = eval_buckets, dic = dic, rev_dic = rev_dic)
 ```
 
-## Inference from model
+Create iterators for training and evaluation datasets.
 
 
-Some helper functions for random sample.
+```R
+vocab <- length(eval_buckets$dic)
 
-```{r}
-cdf <- function(weights) {
-    total <- sum(weights)
-    result <- c()
-    cumsum <- 0
-    for (w in weights) {
-        cumsum <- cumsum+w
-        result <- c(result, cumsum / total)
-    }
-    return (result)
-}
+batch.size = 32
 
-search.val <- function(cdf, x) {
-    l <- 1
-    r <- length(cdf) 
-    while (l <= r) {
-        m <- as.integer((l+r)/2)
-        if (cdf[m] < x) {
-            l <- m+1
-        } else {
-            r <- m-1
-        }
-    }
-    return (l)
-}
+train.data <- mx.io.bucket.iter(buckets = train_buckets$buckets, batch.size = batch.size, 
+                                data.mask.element = 0, shuffle = TRUE)
 
-choice <- function(weights) {
-    cdf.vals <- cdf(as.array(weights))
-    x <- runif(1)
-    idx <- search.val(cdf.vals, x)
-    return (idx)
-}
+eval.data <- mx.io.bucket.iter(buckets = eval_buckets$buckets, batch.size = batch.size,
+                               data.mask.element = 0, shuffle = FALSE)
 ```
 
-We can use random output or fixed output by choosing largest probability.
-
-```{r}
-make.output <- function(prob, sample=FALSE) {
-    if (!sample) {
-        idx <- which.max(as.array(prob))
-    }
-    else {
-        idx <- choice(prob)
-    }
-    return (idx)
-
+## Train the Model
+
+
+This model is a multi-layer RNN for sampling from character-level language models. It has a one-to-one model configuration since for each character, we want to predict the next one. For a sequence of length 100, there are also 100 labels, corresponding the same sequence of characters but offset by a position of +1. The parameters output_last_state is set to TRUE in order to access the state of the RNN cells when performing inference.
+
```

Comment:
Would it be helpful if we mention that currently only one-to-one and many-to-one flavours are supported.

Reply:
it would not be relevant to this tutorial